src/action.ts

```diff
@@ -123,40 +124,86 @@ export async function performAction(
+		saveTorrentFile(tracker, getTag(searchee.name), newMeta);
+		const styledName = chalk.green.bold(newMeta.name);
+		const styledTracker = chalk.bold(tracker);
+		logger.info(`Found ${styledName} on ${styledTracker} - saved`);
+		return SaveResult.SAVED;
+	}
+
+	let destinationDir: string | undefined,
+		ogDownloadDir,
+		linkedFilesRootResult;
+
+	// SET THE CLIENTS SAVE PATH - PREVENTS ERROR IF DATADIR/LINKDIR IS SET
+	// BUT NOT THE TORRENT DATA DIRECTORY AND/OR LINKING THROWS OR IS NOT SET
+	if (searchee.infoHash) {
+		ogDownloadDir = await getClient().getDownloadDir(searchee);
+		destinationDir = ogDownloadDir.isOk()
+			? ogDownloadDir.unwrapOrThrow()
+			: undefined;
+	}
+	if (
+		linkDir &&
+		typeof destinationDir === "string" &&
+		existsSync(destinationDir)
+	) {
+		linkedFilesRootResult = await linkAllFilesInMetafile(
 			searchee,
 			newMeta,
 			tracker,
 			decision
 		);
-		if (linkedFilesRootResult.isErr()) {
-			// TODO
-			return InjectionResult.FAILURE;
-		}
-		linkedFilesRoot = linkedFilesRootResult.unwrapOrThrow();
+		destinationDir = linkedFilesRootResult.isOk()
+			? linkedFilesRootResult.unwrapOrThrow()
+			: destinationDir;
+		/*} else {
+			//linking failed - need to compare file trees or figure out nested linking?
+			//compareFileTrees(newMeta, searchee);
+			//	logInjectionResult(result, tracker, newMeta.name);
+
+			destinationDir = ogDownloadDir.unwrapOrThrow();
+
+			/* extra code for referencing
+				const result =
+					linkedFilesRootResult.unwrapErrOrThrow() as InjectionResult;
+				logInjectionResult(result, tracker, newMeta.name);
+				saveTorrentFile(tracker, getTag(searchee.name), newMeta);
+				return InjectionResult.FAILURE;
+				*/
+	}
+	if (typeof destinationDir !== "string") {
+		logInjectionResult(InjectionResult.FAILURE, tracker, newMeta.name);
+		saveTorrentFile(tracker, getTag(searchee.name), newMeta);
+		return InjectionResult.FAILURE;
 	}
 
-	if (action === Action.INJECT) {
-		// using downloadDir for now but we might want to use linkedFilesRoot instead if we get missing files errors
-		const downloadDir = linkedFilesRoot
-			? dirname(linkedFilesRoot)
-			: linkedFilesRoot;
-		const result = await getClient().inject(newMeta, searchee, downloadDir);
-		logInjectionResult(result, tracker, newMeta.name);
-		if (result === InjectionResult.FAILURE) {
-			saveTorrentFile(tracker, getTag(searchee.name), newMeta);
-		}
-		return result;
-	} else {
+	// NEW compare() for STRICTLY COMPARING FILES (nested linking)
+	const nestedMatch = compareFileTreesIgnoringFolders(newMeta, searchee);
+	const perfectMatch = compareFileTrees(newMeta, searchee);
+
+	// SO SORRY!!!! TERNARY HELL D:
+	const downloadDir =
+		linkedFilesRootResult && linkedFilesRootResult.isOk()
+			? destinationDir
+			: perfectMatch
+			? destinationDir
+			: newMeta.isSingleFileTorrent && nestedMatch
+			? posix.join(destinationDir, searchee.name) // using posix for dev env only. needs to be removed.
+			: undefined;
```

Comment:
ideally this logic lives inside `linkAllFilesInMetafile` and the result is accessible as `linkedFilesRootResult.unwrapOrThrow()`